extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,321 @@
+
+                // Cancel any outstanding requests for this cached fragment
+                cachedEntry.future.cancel(true);
+
+                // Reuse the existing entry
+                cachedEntry.fragmentIndex = fragmentIndex;
+                cachedEntry.future = future;
+            }
+        }
+
+        /**
+         * Will return the {@link CompletableFuture} corresponding to provided fragment index if present in the cache,
+         * else will return {@code null}
+         */
+        @Nullable
+        CompletableFuture<ByteBuffer> getCachedFuture(final int fragmentIndex) {
+            final FragmentContext cachedFragment = bufferCache.get(getIndex(fragmentIndex));
+            if (cachedFragment != null && cachedFragment.fragmentIndex == fragmentIndex) {
+                return cachedFragment.future;
+            }
+            return null;
+        }
+
+        long getSize() {
+            return size;
+        }
+
+        void setSize(final long size) {
+            this.size = size;
+        }
+    }
+
+    private final S3AsyncClient s3AsyncClient;
+    private final String bucket;
+    private final String key;
+    private final int fragmentSize;
+    private final int readAheadCount;
+    private final Duration readTimeout;
+
+    /**
+     * The size of the object in bytes, fetched at the time of first read
+     */
+    private long size;
+    private int numFragmentsInObject;
+
+    private S3ChannelContext context;
+
+    private long position;
+
+    S3SeekableByteChannel(@NotNull final SeekableChannelsProvider.ChannelContext context, @NotNull final URI uri,
+            @NotNull final S3AsyncClient s3AsyncClient, final int fragmentSize, final int readAheadCount,
+            final Duration readTimeout) {
+        final S3Uri s3Uri = s3AsyncClient.utilities().parseUri(uri);
+        this.bucket = s3Uri.bucket().orElse(null);
+        this.key = s3Uri.key().orElse(null);
+        this.s3AsyncClient = s3AsyncClient;
+        this.fragmentSize = fragmentSize;
+        this.readAheadCount = readAheadCount;
+        this.readTimeout = readTimeout;
+        Assert.instanceOf(context, "context", S3ChannelContext.class);
+        this.context = (S3ChannelContext) context;
+        this.size = -1;
+        this.position = 0;
+    }
+
+    @Override
+    public void setContext(@Nullable final SeekableChannelsProvider.ChannelContext context) {
+        // null context equivalent to clearing the context
+        if (context != null && !(context instanceof S3ChannelContext)) {
+            throw new IllegalArgumentException(
+                    "context must be null or an instance of ChannelContext, provided context" +
+                            " of class " + context.getClass().getName());
+        }
+        this.context = (S3ChannelContext) context;
+    }
+
+    @Override
+    public int read(@NotNull final ByteBuffer destination) throws ClosedChannelException {
+        Assert.neqNull(context, "context");
```

Comment:
As discussed verbally, you need to support a `null` context. In that case, you should either perform a single read of the requested size, or a set of concurrent reads for better throughput (adhering to fragment size), with no caching.

Reply:
`null` would be equivalent to clearing the context. I am using `SeekableChannelContext.NULL` to indicate no caching and readahead.